### PR TITLE
Weighted uniform random tipselection + COO heaviest branch tipselection

### PR DESCRIPTION
**Pull request by muXxer**
# Description

This PR adds the Weighted URTS and the COO heaviest branch tipselect for chrysalis step 1.

# ToDo
- [x] Propagate Root txs snapshot indexes in the future cone
- [x] Check the cone for invalid bundles and cache that info
- [x] Question: Do lazy tips remove other tips from the pool?
- [x] Run the solidifier in correct order
- [ ] Solve COO tipselect todos
- [ ] Write test cases

### CODE REVIEW

File: config.json [hunk 1]
```diff
@@ -99,8 +99,11 @@
     }
   },
   "tipsel": {
-    "belowMaxDepthTransactionLimit": 20000,
-    "maxDepth": 5
```
Comment (alexsporn) [inline config.json:103]: why rename maxDepth to belowMaxDepth?

Comment (muXxer) [reply config.json:103]: maxDepth was another thing back then... it was used for maximum starting depth of the random walk.
This belowMaxDepth is the treshold for lazy tips.

File: pkg/model/hornet/transaction_metadata.go [hunk 2]
```diff
@@ -124,6 +127,23 @@ func (m *TransactionMetadata) SetConflicting(conflicting bool) {
...
+	m.Lock()
+	defer m.Unlock()
+
+	m.youngestRootSnapshotIndex = yrtsi
+	m.oldestRootSnapshotIndex = ortsi
+	m.rootSnapshotCalculationIndex = rtsci
+	m.SetModified(true)
+}
+
+func (m *TransactionMetadata) GetRootSnapshotIndexes() (milestone.Index, milestone.Index, milestone.Index) {
```
Comment (alexsporn) [inline pkg/model/hornet/transaction_metadata.go]: maybe give the return values a label?
`(yrtsi milestone.Index, ortsi milestone.Index, rtsci milestone.Index)`

File: pkg/model/mselection/heaviest.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gohornet/hornet/pkg/model/hornet"
+	"github.com/gohornet/hornet/pkg/model/tangle"
+	"github.com/gohornet/hornet/pkg/utils"
+	"github.com/iotaledger/iota.go/trinary"
+	"github.com/willf/bitset"
```
Comment (alexsporn) [inline pkg/model/mselection/heaviest.go]: order the imports

File: pkg/model/mselection/heaviest.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+)
+
+// Errors during milestone selection
+var (
+	ErrWrongReference = errors.New("reference does not match root")
+)
+
+const capIncreaseFactor = 1.15
```
Comment (alexsporn) [inline pkg/model/mselection/heaviest.go]: Comment what this controls

File: plugins/tangle/solidifier.go [hunk 5]
```diff
@@ -383,6 +386,8 @@ func solidifyMilestone(newMilestoneIndex milestone.Index, force bool) {
 
 	conf, err := whiteflag.ConfirmMilestone(cachedMsToSolidify.Retain(), func(tx *tangle.CachedTransaction, index milestone.Index, confTime int64) {
 		Events.TransactionConfirmed.Trigger(tx, index, confTime)
+	}, func(confirmation *whiteflag.Confirmation) {
+		Events.MilestoneConfirmed.Trigger(confirmation)
```
Comment (alexsporn) [inline plugins/tangle/solidifier.go:393]: Move 
```
tangle.SetSolidMilestoneIndex(conf.Index)
Events.SolidMilestoneChanged.Trigger(cachedMsToSolidify) // bundle pass +1
```
to inside this callback

File: pkg/tipselect/urts.go [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+	handler.(func(*Tip))(params[0].(*Tip))
+}
+
+// WalkerStatsCaller is used to signal tip selection events.
+func WalkerStatsCaller(handler interface{}, params ...interface{}) {
+	handler.(func(*TipSelStats))(params[0].(*TipSelStats))
+}
+
+const (
+	// ScoreLazy is a lazy zip and should not be selected.
```
Comment (alexsporn) [inline pkg/tipselect/urts.go]: tip not zip :P

File: pkg/tipselect/urts.go [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+	ScoreLazy Score = iota
+	// ScoreSemiLazy is a somewhat lazy tip.
+	ScoreSemiLazy
```
Comment (alexsporn) [inline pkg/tipselect/urts.go:50]: do you want to call is `ScoreSomewhatLazy`? It sounds weird, but uses the terminology from the RFC

File: pkg/tipselect/urts.go [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+		}
+	}
+}
+
+// removeTipWithoutLocking removes the given tailTxHash from the tipsMap without acquiring the lock.
+func (ts *TipSelector) removeTipWithoutLocking(tailTxHash hornet.Hash) {
+	if tip, exists := ts.tipsMap[string(tailTxHash)]; exists {
+		ts.scoreSum -= int(tip.Score)
+		delete(ts.tipsMap, string(tailTxHash))
+		ts.Events.TipRemoved.Trigger(tip)
```
Comment (alexsporn) [inline pkg/tipselect/urts.go:192]: sure we should trigger an event here while locked?

File: pkg/tipselect/urts.go [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+
+	// iterate over the tipsMap and subtract each tip's score from randScore
+	for _, tip := range ts.tipsMap {
+		// subtract the tip's score from randScore
+		randScore -= int(tip.Score)
+
+		// if randScore reaches zero or below, we return the given tip
+		if randScore <= 0 {
+			// check the "own" score of the tip again to avoid old tips
+			if score := ts.calculateScore(tip.Hash, false, lsmi); score == ScoreLazy {
```
Comment (alexsporn) [inline pkg/tipselect/urts.go]: this is not done in the RFC

Comment (alexsporn) [reply pkg/tipselect/urts.go]: OK, I see you use it to clean up the tip pool

File: pkg/tipselect/urts.go (hunk 7 above)
Comment (alexsporn) [reply pkg/tipselect/urts.go:50]: OK, the RFC mixes up definitions and call it semi lazy later

File: pkg/tipselect/urts.go [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+		cachedApproveeTx.Release(true)
+
+		// if the OTRSI to LSMI delta of the approvee is MaxDeltaTxApproveesOldestRootSnapshotIndexToLSMI, we mark it as such
+		if lsmi-approveeORTSI > ts.maxDeltaTxApproveesOldestRootSnapshotIndexToLSMI {
+			approveesLazy++
+		}
+	}
+
+	// if all available approvees' OTRSI violates the LSMI delta in relation to C2 the tip is lazy too
+	if len(approveeHashes) == approveesLazy {
```
Comment (alexsporn) [inline pkg/tipselect/urts.go:444]: I like you using len here instead of hardcoded 2 as in the RFC, because this way you handle trunk == branch 👍

**Review by alexsporn — APPROVED** [review]

File: pkg/tipselect/urts.go (hunk 8 above)
Comment (muXxer) [reply pkg/tipselect/urts.go:192]: Do you have an idea how to do that, without making everything unreadable and complex? I unfortunately don't have one :(

File: pkg/tipselect/urts.go (hunk 10 above)
Comment (muXxer) [reply pkg/tipselect/urts.go:444]: Exactly, RFC is not optimal in that case :)

File: pkg/whiteflag/white_flag.go [hunk 11]
```diff
@@ -248,11 +273,13 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[string]str
 		}
 
 		if conflicting {
+			wfConf.TailsReferenced = append(wfConf.TailsReferenced, currentTx.GetTxHash())
```
Comment (alexsporn) [inline pkg/whiteflag/white_flag.go]: move this before the if

File: pkg/whiteflag/white_flag.go [hunk 12]
```diff
@@ -248,11 +273,13 @@ func ProcessStack(stack *list.List, wfConf *Confirmation, visited map[string]str
...
 			wfConf.TailsExcludedConflicting = append(wfConf.TailsExcludedConflicting, currentTx.GetTxHash())
 			return nil
 		}
 
 		// mark the given tail to be part of milestone ledger changing tail inclusion set
+		wfConf.TailsReferenced = append(wfConf.TailsReferenced, currentTx.GetTxHash())
```
Comment (alexsporn) [inline pkg/whiteflag/white_flag.go]: if moved before the if, remove 2nd stmt

**Review by alexsporn — APPROVED** [review]

**Review by alexsporn — APPROVED** [review]

**Pull request merged**